synapse/third_party_rules/access_rules.py

```diff
@@ -94,35 +94,56 @@ def on_create_room(self, requester, config, is_requester_admin):
         default rule to the initial state.
         """
         is_direct = config.get("is_direct")
-        rule = None
+        preset = config.get("preset")
+        access_rule = None
+        join_rule = None
 
         # If there's a rules event in the initial state, check if it complies with the
         # spec for im.vector.room.access_rules and deny the request if not.
         for event in config.get("initial_state", []):
             if event["type"] == ACCESS_RULES_TYPE:
-                rule = event["content"].get("rule")
+                access_rule = event["content"].get("rule")
 
                 # Make sure the event has a valid content.
-                if rule is None:
+                if access_rule is None:
                     raise SynapseError(400, "Invalid access rule")
 
                 # Make sure the rule name is valid.
-                if rule not in VALID_ACCESS_RULES:
+                if access_rule not in VALID_ACCESS_RULES:
                     raise SynapseError(400, "Invalid access rule")
 
                 # Make sure the rule is "direct" if the room is a direct chat.
                 if (
-                    (is_direct and rule != ACCESS_RULE_DIRECT)
-                    or (rule == ACCESS_RULE_DIRECT and not is_direct)
+                    (is_direct and access_rule != ACCESS_RULE_DIRECT)
+                    or (access_rule == ACCESS_RULE_DIRECT and not is_direct)
                 ):
                     raise SynapseError(400, "Invalid access rule")
 
-        # If there's no rules event in the initial state, create one with the default
-        # setting.
-        if not rule:
+            if event["type"] == EventTypes.JoinRules:
+                join_rule = event["content"].get("join_rule")
+
+        if access_rule:
+            # If there's an access rules event in the initial state, check if the prefix
+            # or the join rule in use is compatible (i.e. if it involves a "public" join
+            # rule, the access rule must be "restricted"). We don't need to check that if
+            # there's no access rule provided, as in this case the access rule will
+            # default to "restricted", with which any join rule is allowed.
```

Comment:
I would move this after we've handled the case of no `access_rule`, so that we're checking what we are *actually* defaulting the rule to rather than what we think we will (especially if we end up changing things in the future and forgetting to update this branch).

In particular, this comment isn't actually true in the case where `is_direct=True`, which is probably not a problem exactly, but still.